design/design-doc-composition.md

```diff
@@ -0,0 +1,1400 @@
+This document proposes that all contemporary resource claims and classes be
+marked as deprecated once support for composition has been added to Crossplane.
+
+The class and claim model requires that managed resources include a handful of
+fields that are defunct under the design proposed by this document:
+
+* `spec.classRef` is superseded by `spec.compositionRef`, which exists at the
+  composite resource rather than managed resource level.
+* `spec.claimRef` is superseded by `spec.requirementRef`, which exists at the
+  composite resource rather than managed resource level.
+* `status.bindingPhase` no longer exists. Any extant managed resource with a
+  composite resource as a controller reference is inherently 'bound' to that
+  composite resource. A composite resource is bound to a resource requirement
+  when each references the other.
+
+The contemporary resource claim controller uses the `status.bindingPhase` of an
+existing managed resource to determine whether it may be claimed by a resource
+claim that explicitly specifies said managed resource as its `spec.resourceRef`.
+A composed managed resource would thus appear to the resource claim controller
+to be available for binding due to their unset `status.bindingPhase`. This can
+be avoided by updating the resource claim controller to disallow resource claims
+from binding to a managed resource with a controller reference, as composed
+managed resources will specify their composite resource as their controller
+reference.
+
+The semantics of the managed resource `spec.reclaimPolicy` also change under the
+proposed design. In the contemporary class and claim model `spec.reclaimPolicy`
+is semantically identical to the [reclaim policy of persistent volumes]. It
+controls:
+
+* Whether the external resource is deleted when the managed resource is deleted.
+* Whether the managed resource is deleted when the resource claim it is bound to
+  is deleted.
+
+Under the proposed design `spec.reclaimPolicy` exists at two levels:
+
+* At the managed resource level the reclaim policy determines what happens to
+  an external resource when its corresponding managed resource is deleted.
+* At the composite resource level the reclaim policy determines what happens to
+  the composite resource when it is bound to a resource requirement that is
+  deleted. There is no `spec.bindingPhase`, so a 'retained' composition becomes
+  available to subsequent resource requirements.
+
+This document recommends these fields be renamed once resource claims are
+removed in order to avoid any implication that they function identically to the
+reclaim policy of a persistent volume.
+
+### For Applications
+
+Application composition introduces new resource kinds (e.g. `Composition` and
+`ApplicationDefinition`) that do not affect existing applications built using
+the  templating-controller (aka "Template Stacks"). This document proposes that
```

Comment:
I am suspicious about deprecating template stacks. The differences between the composition and template stacks are:
* Template stacks have a file system and the authoring experience is better as you can create your own folder structure instead of having to put everything under one YAML.
* Makes it possible to package a set of YAMLs as Crossplane Package for delivery.

IMO, we'll need to capture these advantages with Composition stuff one way or another before deprecation.

Reply:
> IMO, we'll need to capture these advantages with Composition stuff one way or another before deprecation.

Agreed - I don't think templating-controller will go anywhere until we have something like a `HelmComposition` that includes all or most of its functionality. I do think we should aim to bring these two closer together though given how similar they are.